File: nodejs/flightradarapi/entities/entity.js
class Entity {
    /**
     * Representation of a real entity, at some location.
     *
     * @param {number} latitude
     * @param {number} longitude
     */
    constructor(latitude=null, longitude=null) {
        this.__set_position(latitude, longitude);
    }

    __set_position(latitude, longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    __createGetterMethodFor(target, recursive = true) {
        function getter(x, defaultValue = null) {
            let value = defaultValue;

            if (this.hasOwnProperty(x)) {
                value = this[x];
            }

            if (recursive && (value != null && typeof(value) === "object")) {
                value.get = getter;
            }
            return value;
        }

        target.get = getter;
        return target;
    }

    __getDetails(data) {
        if (data == null) {
            data = {};
        }
        return this.__createGetterMethodFor(data);
    }

    __getInfo(info) {
        return (info || info === 0) && (info !== this.__default_text) ? info : this.__default_text;
    }

    /**
     * Return the distance from another entity (in kilometers).
     *
     * @param {Entity} entity
     * @returns {number}
     */
    get_distance_from(entity) {
        let lat1 = Math.radians(this.latitude);
        let lon1 = Math.radians(this.longitude);

        let lat2 = Math.radians(entity.latitude);
        let lon2 = Math.radians(entity.longitude);

        return Math.acos(
            Math.sin(lat1) * Math.sin(lat2) +
            Math.cos(lat1) * Math.cos(lat2) * Math.cos(lon2 - lon1)
        ) * 6371;
    }
}

module.exports = Entity;

File: python/FlightRadarAPI/entities/entity.py
# -*- coding: utf-8 -*-

from math import acos, cos, radians, sin
from typing import Any, Optional

EARTH_RADIUS_KM = 6371


class Entity:
    """
    Representation of a real entity, at some location.
    """

    _default_text = "N/A"

    def __init__(self, latitude: Optional[float], longitude: Optional[float]):
        """
        Constructor of the Entity class.
        """
        self._set_position(latitude, longitude)

    def _set_position(self, latitude: Optional[float], longitude: Optional[float]) -> None:
        self.latitude = latitude
        self.longitude = longitude

    def _get_info(self, info: Any, default: Optional[Any] = None) -> Any:
        """
        Return `info` unless it is None or the sentinel for missing values; in that
        case return `default` (or the entity's default sentinel).
        """
        default = default if default is not None else self._default_text
        return info if info is not None and info != self._default_text else default

    def get_distance_from(self, entity: "Entity") -> float:
        """
        Return the distance from another entity (in kilometers).
        """
        if (self.latitude is None or self.longitude is None
                or entity.latitude is None or entity.longitude is None):
            raise ValueError("Cannot calculate distance: one or both entities have no position.")

        lat1, lon1 = radians(self.latitude), radians(self.longitude)
        lat2, lon2 = radians(entity.latitude), radians(entity.longitude)

        return acos(sin(lat1) * sin(lat2) + cos(lat1) * cos(lat2) * cos(lon2 - lon1)) * EARTH_RADIUS_KM
